### PR TITLE
feat(payments): Add audit events for PaymentUpdate update

### LINKED ISSUE #4682: Add audit events for PaymentUpdate update
Created from #4525 

This covers adding events for PaymentUpdate operation

This event should include the payment data similar to [PaymentCancel](https://github.com/juspay/hyperswitch/pull/4166)

It should also include any metadata for the event  
e.g reason for payment rejection, error codes, rejection source etc

### Submission Process:
- Ask the maintainers for assignment of the issue, you can request for assignment by commenting on the issue itself.
- Once assigned, submit a pull request (PR).
- Maintainers will review and provide feedback, if any. 
- Maintainers can unassign issues due to inactivity, [read more here](https://github.com/juspay/hyperswitch/wiki/Hacktoberfest-Contribution-Rules).

Refer [here](https://github.com/juspay/hyperswitch/blob/main/docs/TERMS_OF_CONTEST.md) for Terms and conditions for the contest.

### PR DESCRIPTION
**Pull request by Rutam21**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR adds an Audit event for the PaymentUpdate operation.

### Additional Changes

- [x] This PR modifies the files below

- `crates/router/src/core/payments/operations/payment_update.rs`
- `crates/router/src/events/audit_events.rs`

## Motivation and Context

This PR fixes #4682 

## How did you test it
- Hit the `Payments - Create` endpoint.
Make a payment with `"confirm": false`
```bash
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_LZcIA7XeMpnK5satp4CDvjcnHCaeKTBcosyuBBJaknZu1odhFo96cwS0nSdfzuJF' \
--data-raw '{
    "amount": 6540,
    "currency": "USD",
    "confirm": false,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 6540,
    "customer_id": "StripeCustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+65",
    "description": "Its my first payment request",
    "authentication_type": "no_three_ds",
    "return_url": "https://google.com",
    "payment_method": "card",
    "payment_method_type": "credit",
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "joseph Doe",
            "card_cvc": "123"
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "8056594427",
            "country_code": "+91"
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "udf1": "value1",
        "new_customer": "true",
        "login_date": "2019-09-10T10:11:12Z"
    },
    "profile_id": "pro_v5sFoHe80OeiUlIonocM"
}'
```
- Now hit the `Payments - Update` endpoint.

```bash
{
  "amount": 20000,
  "currency": "SGD",
  "confirm": false,
  "capture_method": "automatic",
  "capture_on": "2022-09-10T10:11:12Z",
  "amount_to_capture": 20000,
  "email": "joseph@example.com",
  "name": "joseph Doe",
  "phone": "8888888888",
  "phone_country_code": "+65",
  "description": "Its my first payment request",
  "authentication_type": "no_three_ds",
  "payment_method": "card",
  "return_url": "https://google.com",
  "payment_method_data": {
    "card": {
      "card_number": "4242424242424242",
      "card_exp_month": "10",
      "card_exp_year": "25",
      "card_holder_name": "joseph Doe",
      "card_cvc": "123"
    }
  },
   "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US"
        }
    },
  "statement_descriptor_name": "joseph",
  "statement_descriptor_suffix": "JS",
  "metadata": {
    "udf1": "value1",
    "new_customer": "true",
    "login_date": "2019-09-10T10:11:12Z"
  }
}
```

Audit log will get generated and can be viewed in hyperswitch-audit-events topic in kafka with `PaymentUpdate` event-type.
Amount parameter can be verified in the event.
Screenshot 2024-11-07 at 12 42 34 PM
Screenshot 2024-11-07 at 12 42 42 PM

Now payment can be completed by hitting the `Payments - Confirm` endpoint.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code

### CODE REVIEW

**Comment by Rutam21** [comment]
Please review and suggest changes, if any.

cc: @Isampras @ShivanshMathurJuspay @ivor11

**Comment by lsampras** [comment]
Hey @Rutam21 ,
Can you share the event json or a screenshot of the event,
Can refer to pr https://github.com/juspay/hyperswitch/pull/4166 for reference.

**Comment by Rutam21** [comment]
@lsampras I checked #4166, can you please point me out how I can get a event screenhsot or event json for this update? It seems like you were using something similar to Postman to hit the API endpoint. If you could please let me know the steps, I would follow along and update.

**Comment by maverox** [comment]
> @lsampras I checked #4166, can you please point me out how I can get a event screenhsot or event json for this update? It seems like you were using something similar to Postman to hit the API endpoint. If you could please let me know the steps, I would follow along and update.

For the UI in this screenshot, you can find the docs to set up Kafka, kafka-ui (the interface in the screenshot) and ClickHouse using Docker Compose here: https://github.com/juspay/hyperswitch/tree/main/crates/analytics/docs

and then try generating said events and monitor logs in kafka-ui.

Also the docs are self-explanatory but still if in any doubt ping us in [this](https://hyperswitch-io.slack.com/archives/C04JXRZ338A) slack channel

**Comment by gorakhnathy7** [comment]
Hey @Rutam21 Any updates.

**Comment by Rutam21** [comment]
@maverox I tried updating the respective `toml` files and then starting the app using docker-compose. But this errors out while starting the `hyperswitch-server-1`. Please let me know if there are some additional settings that I am missing here. Thanks.

Screenshot 2024-10-31 050533

cc: @gorakhnathy7

**Comment by maverox** [comment]
> @maverox I tried updating the respective `toml` files and then starting the app using docker-compose. But this errors out while starting the `hyperswitch-server-1`. Please let me know if there are some additional settings that I am missing here. Thanks.
> 
> Screenshot 2024-10-31 050533
> 
> cc: @gorakhnathy7

hey bro this happens with me as well when i am running the server locally (not containerized) 
basically if you do 
```bash 
cargo r
```

so if your server is running and all these other containers are running as well there shouldn't be any problem try
going to any of the analytics services
kafka-ui or clickhouse playground

make some payments using the api or dashboard and check for new logs there.

**Review by tsdk02 — APPROVED** [review]

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 1]
```diff
@@ -919,6 +920,11 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
             )
             .await
             .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+        req_state
+            .event_context
+            .event(AuditEvent::new(AuditEventType::PaymentUpdate))
+            .with(payment_data.to_event())
+            .emit();
```
Comment (tsdk02) [inline crates/router/src/core/payments/operations/payment_update.rs:927]: Can you send `amount` to be logged here
You can take reference from this PR: https://github.com/juspay/hyperswitch/pull/4913/files

File: crates/router/src/events/audit_events.rs [hunk 2]
```diff
@@ -27,6 +27,7 @@ pub enum AuditEventType {
         capture_amount: Option<MinorUnit>,
         multiple_capture_count: Option<i16>,
     },
+    PaymentUpdate,
```
Comment (tsdk02) [inline crates/router/src/events/audit_events.rs:30]: Add `amount` here as well.

**Comment by Rutam21** [comment]
@tsdk02 I will push another commit to add the `amount` parameter. Thanks for the suggestion.

**Comment by tsdk02** [comment]
Hey @Rutam21, a gentle reminder to add the suggested changes to this PR, so it can be merged.

File: crates/router/src/core/payments/operations/payment_update.rs (hunk 1 above)
Comment (Rutam21) [reply crates/router/src/core/payments/operations/payment_update.rs]: I have added the amount in the fresh commit.

**Comment by Rutam21** [comment]
@tsdk02 I have added the `amount` parameter to the audit event. Let me know if this looks good. Thanks.

File: crates/router/src/core/payments/operations/payment_update.rs [hunk 3]
```diff
@@ -925,6 +926,12 @@ impl<F: Clone> UpdateTracker<F, PaymentData<F>, api::PaymentsRequest> for Paymen
             )
             .await
             .to_not_found_response(errors::ApiErrorResponse::PaymentNotFound)?;
+        let amount = payment_data.amount.into();
```
Comment (tsdk02) [inline crates/router/src/core/payments/operations/payment_update.rs:929]:
```suggestion
        let amount = payment_data.amount;
```

File: crates/router/src/events/audit_events.rs [hunk 4]
```diff
@@ -27,6 +27,9 @@ pub enum AuditEventType {
         capture_amount: Option<MinorUnit>,
         multiple_capture_count: Option<i16>,
     },
+    PaymentUpdate {
+        amount: Option<MinorUnit>,
```
Comment (tsdk02) [inline crates/router/src/events/audit_events.rs:31]:
```suggestion
        amount: Amount,
```

Comment (tsdk02) [reply crates/router/src/events/audit_events.rs:31]: U can import 
```bash
use api_models::payments::Amount;
```
in this file

File: crates/router/src/core/payments/operations/payment_update.rs (hunk 3 above)
Comment (Rutam21) [reply crates/router/src/core/payments/operations/payment_update.rs]: This change is applied now.

File: crates/router/src/events/audit_events.rs (hunk 4 above)
Comment (Rutam21) [reply crates/router/src/events/audit_events.rs]: Thanks. The change is applied now.

**Comment by Rutam21** [comment]
@tsdk02 Please review and suggest changes, if any. Thanks.

**Review by tsdk02 — APPROVED** [review]

**Review by Chethan-rao — DISMISSED** [review]

**Comment by Rutam21** [comment]
The merge conflict is resolved and the branch is updated.

Please review, approve and add to the merge queue. Thanks.

cc: @Chethan-rao @tsdk02 @likhinbopanna

**Review by tsdk02 — APPROVED** [review]

**Review by Chethan-rao — APPROVED** [review]

**Pull request merged**